### DISCUSSION TITLE
New side project, having UNO problems..

### DISCUSSION THREAD

**Discussion by MickTheMechanic** [ImpulseAdventure/GUIslice#361] [Q&A]
Heres my side project, Its a portable 12v compressor I threw together using an old air suspension compressor from a BMW and a couple of empty oxygen tanks. It will be perfect to take to motorsport events along with my pyrometer.

![IMG_0517](https://user-images.githubusercontent.com/44751666/110371673-983e7800-804d-11eb-921e-5f4112169989.jpeg)
![IMG_0518](https://user-images.githubusercontent.com/44751666/110371687-9bd1ff00-804d-11eb-80fb-e65e7efae6fc.jpeg)
![IMG_0510](https://user-images.githubusercontent.com/44751666/110371708-a1c7e000-804d-11eb-883f-935832a13295.jpeg)

Its supposed to be a quick project using parts I have lying around, I have a spare UNO and the old screen from my pyromter, so I want to use these to display the pressure and switch a relay for the compressor on and off according to pressure inside the tanks.

Unfortunately I cant get xRingGauge to work on the UNO. Is it actually possible to use xRingGauge with the UNO's limited resources?

**Comment by MickTheMechanic** [comment]
I have managed to build a working example, but I have run out of space (only one ring gauge and one text element!) and there are a few more minor things I would like to do. I am wondering if its possible to gain some more memory somehow, or if I should just bite the bullet and buy another ESP32 board.

Here is the sketch and config file:
[2Compressor.zip](https://github.com/ImpulseAdventure/GUIslice/files/6111734/2Compressor.zip)

[ard-shld-mcufriend.zip](https://github.com/ImpulseAdventure/GUIslice/files/6111739/ard-shld-mcufriend.zip)

It would be really awesome if one of you guys could have a quick look and see if there is some was to squeeze a bit more memory out of it. If its not possible then Ill order an ESP.
Thanks :)

**Reply by ImpulseAdventure** [reply]
Hey Mick -- your compressor looks great!

The UNO is extremely limited in resources, so incorporating GUI elements like ring gauges will be extremely tight on memory. Nonetheless, if you'd like to squeeze a bit more capability, there are some things we can do:
- It looks like your primary issue is with the UNO's limited program memory, not RAM. In that case, moving elements from RAM to Flash won't really help. Instead, the largest gain would likely come from reducing the cost of the trigonometry functions used by Ring Gauge.
- The Ring Gauge uses a lookup table for fixed-point sine/cosine functions. I have attached a modified version of GUIslice that reduces the lookup table from 256-entry to 128-entry or 64-entry. If you add `#define LUT_SIN_64` to your config file, you should save 384B of PROGMEM while still maintaining acceptable rendering quality on smaller TFTs. 

I have attached the modified GUIslice.c here. Note that it is based on the current **master** branch of GUIslice (not the Int384-Input2* branches) as it sounds like your compressor sketch wouldn't be using the external inputs.
[GUIslice-disc361.zip](https://github.com/ImpulseAdventure/GUIslice/files/6117851/GUIslice-disc361.zip)

But ultimately I would strongly recommend either getting a cheap ATmega2560 (which has a similar pinout but lager footprint) or going to a more capable MCU, as undoubtedly you'll want to add some additional functionality :)

Thanks
